no more empty cell added at the end of a row

tschoepping opened the pull request:
When reading a CSV file, no more empty cells will be added at the end of a row. As a result, the reported column count should now be correct.

d99kris commented:
Thanks for the PR! I've been kind of busy the last week, so I haven't had time to look at it yet. I notice that the Appveyor (Windows) build/unittests failed. It would be good to have it fixed before I review the PR.

d99kris commented:
I just noticed that #37 and #38 contain a commit from here. Perhaps we should close #36 and #37 and just track the final PR under #38?

tschoepping commented:
#36, #37 and #38 are actually three distinct changes, but I realized too late that #37 and #38 contain #36. Anyway, those are three independent features, which you can review separately. In the latest commit on my [master](https://github.com/tschoepping/rapidcsv), those are already combined.

I'll have a look at the failed checks next week.

d99kris commented:
Ok noted (sorry I hadn't looked at the details). 
A heads-up - there was a rather big PR #35 submitted just before your PRs, which I intend to merge first (because it was submitted first). This may or may not affect your PRs (hopefully not), just wanted to let you know (in case you want to wait for that to be merged before any fix-ups).
Yep fixing the unit test failures would be good. And also adding some test, or adding some Github issue describing what fails and the suggestion for how it should work could be good.

tschoepping commented:
My little mistake is making those PRs quite dirty - sorry for that. I merged the fix of #37 to my master, so here we go. Commit 59e192e should still be one to consider for this PR.

wingunder commented:
Hi @tschoepping,

I tried your patch (merged with `d99kris/master`), and found that it does not correctly handle empty columns at the end of a data set. It is reproducible with the following code. By commenting the `#define BREAK_TEST` line, things work, but as soon as `BREAK_TEST` is defined, it breaks.

Correct me if I'm wrong, but I think that the csv string is legitimate in both cases.

```c++
#include <rapidcsv.h>

int main()
{
  const std::string csv =
    "Id,Name\n"
    "0,John\n"
#define BREAK_TEST
#ifdef BREAK_TEST
    "1,\n"
#else
    "1,Mary\n"
#endif
    ;
  try
  {
    std::stringstream sstream(csv);
    rapidcsv::Document doc(sstream);
    const auto avgAge = doc.GetColumn<std::string>("Name");
  }
  catch (const std::exception& ex)
  {
    std::cout << ex.what() << std::endl;
    return 1;
  }
  return 0;
}
```

d99kris commented:
Good catch regarding reading the empty cell @wingunder - I too believe that is a valid CSV cell.

I'm also wondering whether the single quote handling perhaps could be done in a separate PR/discussion? Then this PR could cover just adding support for line-breaks within quotes, and maybe be renamed to something like "Add support for quoted line-breaks in cells".

I am a little hesitant to add the single quote handling. I somehow feel it would be sufficient to handle one type of quoting, but I could be open to make it configurable by the user (so they can choose when to quote using " or ' or something else). Single-quotes are tricky to support by default as they are also used for other purposes than quoting.

tschoepping commented:
I did some brief research on actual specification of CSV, but I could not find a final answer on how trailing empty cells _must_ be handled. Actually I am now quite convinced, that the basic idea of a document having a specific width and height might be not quite correct in the first place. Even though the file only specified values for n rows an m columns, accessing element (n+1,m+1) should still be valid and return an empty string, imho.

Furthermore, have a look at this [example.csv](https://github.com/d99kris/rapidcsv/files/3908228/example.txt). What is the actual width of this document? Or are files like this not valid at all?

This PR can also be called a performance optimization, though. I see nor reason, why trailing empty cells need to be stored and occupy memory. The performance impact, however, probably is negligibly small, though.

So, why do I need this "dropping" of empty trailing cells in the first place? I need my files to specify quadratic matrices and as a first sanity check, I test for the number of rows being equal to the number of columns. In some files, however, there is an additional separator after the last value in a line, just before the line break. As a result, the data seems to be too wide, which they are arguably not. So with this PR, trailing separators, which are immediately followed by a line break, are ignored and no additional cell is added to that row in the document object.

d99kris commented:
Hi @tschoepping - thanks a lot for providing the detailed background and motivation for the PR, it's very helpful!

> Furthermore, have a look at this example.csv. What is the actual width of this document? Or are files like this not valid at all?

I _believe_ it's a valid CSV document, but yeah it's not clear what the actual width is.

I'm still a bit hesitant to merge this PR. The reason is my own original use-case for rapidcsv. I used it to parse financial time-series data retrieved from the web. In my experience such data always have a fixed number of cells (or commas), but occasionally the actual cell value would be missing, so the data could look something like:
```
    Open,High,Low,Close,Volume,Adj Close
    64.529999,64.800003,64.139999,64.620003,21705200,64.620003
    64.419998,,64.190002,64.620003,20235200,64.620003
    64.330002,64.389999,64.050003,64.360001,19259700,
    64.610001,64.949997,64.449997,64.489998,19384900,64.489998
    64.470001,64.690002,64.300003,64.620003,,64.620003
```
And the data would generally be consumed/parsed on a column-basis, like getting all values from column "Volume" (and then I'd use rapidcsv's CustomConverter to default missing values to NaN, zero or something else suitable).

I do believe this could be a more common use-case than the variable number of cells. Soo, I guess what I would consider to encompass your request would be to introduce an option, maybe in SeparatorParams (perhaps called IgnoreEmptyCells, DropEmptyCells or something), that enables the dropping of empty cells. The option would need to be disabled/false by default, to prevent it from breaking other usages.

If you'd like to pursue the merge request, please help introduce such an option. It would also be very useful if a test was added to test it, but if it's tricky I can help to do that (when I find time). 

Finally, I would prefer if only a single issue is addressed by each PR, so please leave out the other changes from this PR and just let this PR cover the dropping of empty cells. Alternatively we'll just wait with merging this PR until the changes in your other PR is merged.

Sorry for being a bit picky, and thanks for helping to contribute.

tschoepping commented:
> I would prefer if only a single issue is addressed by each PR, so please leave out the other changes from this PR and just let this PR cover the dropping of empty cells.

I reverted the merges from my master, so it's only about the dropping of empty trailing cells now. I kept that messy history though, as I'm not sure what GitHub will do when I force-edit my commit history.

I just introduced an additional flag to the SeparatorParams struct, which allows to enable the dropping mechanism. It is disabled by default, as you asked. However, I decided to put the additional constructor argument `pDropEmptyTrailingCells` between `pTrim` and `pHasCR`, so existing code, which intends to set `pHasCR` will now set `pDropEmptyTrailingCells` instead and thus behavior will not be as expected anymore. I leave it up to you, whether you prefer the more sensible order of arguments or reorder them to keep compatibility with existing code. 

Unfortunately, there is a conflict in a comment section, but I cannot see the reason for this =/

Furthermore, I'd like to note that I haven't added any tests to verify the introduced feature yet. I can do so as soon as you have decided about the details of the implementation ;)

d99kris commented:
Thanks for making the requested changes. I still feel a bit uneasy about it though. Why would we drop just last cells on a row that are empty? Vs. all empty cells?

I'm starting to feel that this functionality should maybe not be part of rapidcsv, it doesn't seem generic enough to be part of this type of library.. Do you know any other CSV reading software that treats the data as you suggest rapidcsv to do?

d99kris commented:
Sorry, I should've highlighted my concerns earlier, but I think I didn't properly understand the original suggested change scope as the PR contained a bit of other changes.

tschoepping commented:
> Why would we drop just last cells on a row that are empty? Vs. all empty cells?

Since data is stored row-wise, you can apply a simple heuristic: If the selected column is beyond the actual length of the selected row (in the internal data structure), the cell must be empty. As a result you could store an empty row as a vector of length 0 and still validly access all cells of that row ("nonexistent" cells are handled like empty ones).
Dropping cells somewhere _within_ a row (i.e. there are cells that are not empty "to the right"), however, would be more complicated and probably not worth the effort, as you need some sparse matrix format.

Anyway, my PR shall primarily tackle two cases:
1) The CSV file defines a m x n matrix, but the the last k columns contain only empty cells. In this case the matrix would be stored as m x (n-k) matrix.
2) Lengths of individual rows in the CSV file differ. In this case there is no overall matrix size and each row is cropped to its actual length.
 
> Do you know any other CSV reading software that treats the data as you suggest rapidcsv to do?

I'm not very familiar with a lot of CSV software. I just searched for a library I could easily use for my project and rapidcsv was very close ;)

> I'm starting to feel that this functionality should maybe not be part of rapidcsv

Your project, your decision ;) I'm absolutely fine in either case.

d99kris commented:
I've contemplated this a bit and decided not to accept this PR. Thanks anyway for the contribution and input.
